test/ApplicationTest.php

```diff
@@ -621,4 +625,66 @@ public function testPipingNotInvokableErrorMiddlewareRisesException()
         $this->setExpectedException(InvalidMiddlewareException::class);
         $handler('foo', 'bar', 'baz', 'bat');
     }
+
+    public function invalidRequestExceptions()
+    {
+        return [
+            'invalid file'             => [
+                InvalidArgumentException::class,
+                'Invalid value in files specification',
+            ],
+            'invalid protocol version' => [
+                UnexpectedValueException::class,
+                'Unrecognized protocol version (foo-bar)',
+            ],
+        ];
+    }
+
+    /**
+     * @runInSeparateProcess
+     * @preserveGlobalState disabled
+     * @dataProvider invalidRequestExceptions
+     * @param string $expectedException
+     * @param string $message
+     */
+    public function testRunInvokesFinalHandlerWhenServerRequestFactoryRaisesException(
+        $expectedException,
+        $message
+    ) {
+        // try/catch is necessary in the case that the test fails.
+        // Assertion exceptions raised inside prophecy expectations normally
+        // are fine, but in the context of runInSeparateProcess, these
+        // lead to closure serialization errors. try/catch allows us to
+        // catch those and provide failure assertions.
+        try {
+            $requestFactory = Mockery::mock('alias:' . ServerRequestFactory::class)
+                ->shouldReceive('fromGlobals')
+                ->withNoArgs()
+                ->andThrow($expectedException, $message)
+                ->once()
+                ->getMock();
+
+            $finalHandler = function ($request, $response, $err = null) use ($expectedException, $message) {
+                $this->assertEquals(400, $response->getStatusCode());
+                $this->assertInstanceOf($expectedException, $err);
+                $this->assertEquals($message, $err->getMessage());
+                return $response;
+            };
+
+            $emitter = $this->prophesize(EmitterInterface::class);
+            $emitter->emit(Argument::that(function ($response) {
+                $this->assertInstanceOf(ResponseInterface::class, $response, 'Emitter did not receive a response');
+                $this->assertEquals(400, $response->getStatusCode());
+                return true;
+            }))->shouldBeCalled();
+
+            $app = new Application($this->router->reveal(), null, $finalHandler, $emitter->reveal());
+
+            $app->run();
+        } catch (\Throwable $e) {
+            $this->fail($e->getMessage());
+        } catch (\Exception $e) {
```

Comment:
The `Zend\Expressive\Exception` is already imported, so we need to have this one globally qualified. I'll import `Throwable`, though.